### PR TITLE
Publish corefx test assets

### PR DESCRIPTION
**Pull request by mikem8361**
So coreclr and other repos can run the corefx tests.

New step to publish the test zip files to dotnet-core blob storage.

Publishes the build manifest from above to blob storage path:

https://dotnetfeed.blob.core.windows.net/dotnet-core/corefx-tests/$(Version)/$(TargetOS).$(ArchGroup)/$(TargetGroup)/corefx-test-assets.xml

Declare DotNet-Blob-Feed only in official builds

### CODE REVIEW

File: eng/pipelines/helix.yml [hunk 1]
```diff
@@ -12,6 +12,7 @@ parameters:
   officialBuildId: ''
   enableAzurePipelinesReporter: '' # true | false
   outerloop: '' # true | false
+  buildExtraArguments: ''
```
Comment (safern) [inline eng/pipelines/helix.yml]: Is this a leftOver?

File: eng/pipelines/corefx-base.yml [hunk 2]
```diff
@@ -172,6 +184,25 @@ jobs:
           - ${{ if ne(job.customBuildSteps[0], '') }}:
             - ${{ job.customBuildSteps }}
 
+          - script: $(_msbuildCommand) eng/publish.proj
+                  /t:PublishTestAssets
+                  /p:FilesToPublishPattern=$(Build.SourcesDirectory)/artifacts/helix/**/*.zip
```
Comment (safern) [inline eng/pipelines/corefx-base.yml]: Should this only be executed if submitToHelix == true? There will be no .zip files if submitToHelix is not true.

Also, should we only do this for official builds now?

File: eng/pipelines/helix.yml [hunk 3]
```diff
@@ -29,6 +30,7 @@ steps:
             /p:IsExternal=${{ parameters.isExternal }}
             /p:Creator=${{ parameters.creator }}
             /p:OfficialBuildId=${{ parameters.officialBuildId }}
+            /p:ContinuousIntegrationBuild=true
```
Comment (safern) [inline eng/pipelines/helix.yml]: Why is this needed at all?

File: eng/publish.proj [hunk 4]
```diff
@@ -104,4 +104,56 @@
...
\ No newline at end of file
+
+  <Target Name="PublishTestAssets">
+    <Error Text="The ExpectedFeedUrl property wasn't supplied." Condition="'$(ExpectedFeedUrl)' == ''" />
+    <Error Text="The AccountKey property wasn't supplied." Condition="'$(AccountKey)' == ''" />
+    <Error Text="FilesToPublishPattern property wasn't supplied." Condition="'$(FilesToPublishPattern)' == ''" />
+
+    <PropertyGroup>
+      <_TestAssetVersion>$(PackageVersion)-$(VersionSuffix)</_TestAssetVersion>
+      <_FileRelativePathBase>corefx-tests/$(_TestAssetVersion)/$(OSGroup).$(ArchGroup)/$(TargetGroup)/</_FileRelativePathBase>
```
Comment (safern) [inline eng/publish.proj]: Is TargetGroup needed? Will another repo run corefx tests targeting uap/netfx?

File: eng/pipelines/corefx-base.yml (hunk 2 above)
Comment (safern) [reply eng/pipelines/corefx-base.yml]: I think since this needs DotNet-Blob-Feed -- it should only be published in official builds, right?

File: eng/pipelines/helix.yml (hunk 1 above)
Comment (mikem8361) [reply eng/pipelines/helix.yml]: Yes, I'll remove.

File: eng/pipelines/corefx-base.yml (hunk 2 above)
Comment (mikem8361) [reply eng/pipelines/corefx-base.yml]: I'll add a if submitToHelix.  

_publishTests should only be true in official builds, but I'll also add an if officialbuild

File: eng/pipelines/helix.yml (hunk 3 above)
Comment (mikem8361) [reply eng/pipelines/helix.yml]: This is needed so Arcade correctly generates the version build variables that I use in the blob storage name: $(PackageVersion) and $(VersionSuffix)

File: eng/publish.proj (hunk 4 above)
Comment (mikem8361) [reply eng/publish.proj]: It is for the future. Currently none of the uap/netfx jobs publish tests but we may want them in the future and I assume (maybe incorrectly) that they are built differently than netcoreapp framework.

File: eng/pipelines/corefx-base.yml (hunk 2 above)
Comment (safern) [reply eng/pipelines/corefx-base.yml]: Adding the official build and submitToHelix if, will not pollute the build UI log. So it will avoid having skipped steps in our PR builds.

File: eng/pipelines/helix.yml (hunk 3 above)
Comment (safern) [reply eng/pipelines/helix.yml]: Really, when sending jobs to helix? I didn't see any change in sendtohelix.proj that uses those variables. I did actually see it for publish.proj, which makes sense.

File: eng/pipelines/linux.yml [hunk 5]
```diff
@@ -28,6 +28,7 @@ jobs:
             _dockerContainer: rhel7_container
             _buildScriptPrefix: ''
             _buildExtraArguments: ''
+            _publishTests: true
```
Comment (safern) [inline eng/pipelines/linux.yml]: This is used in both, official build and PR builds, so that is another reason why we need to if condition that step for official builds.

File: eng/pipelines/helix.yml (hunk 3 above)
Comment (mikem8361) [reply eng/pipelines/helix.yml]: Sorry, you are right. It isn't needed here anymore.

File: eng/pipelines/linux.yml (hunk 5 above)
Comment (mikem8361) [reply eng/pipelines/linux.yml]: Yes, I've added both conditions (submitToHelix/isOfficialBuild

File: eng/pipelines/helix.yml (hunk 3 above)
Comment (safern) [reply eng/pipelines/helix.yml]: Btw, I opened: https://github.com/dotnet/arcade/issues/2073 since -ci flag should set that msbuild property.

File: eng/pipelines/corefx-base.yml [hunk 6]
```diff
@@ -55,6 +59,11 @@ jobs:
     - template: ../common/templates/job/job.yml
       parameters:
         variables:
+        - name: _dotnetFeedUrl
+          value: https://dotnetfeed.blob.core.windows.net/dotnet-core/index.json
+        - name: _targetOS
```
Comment (safern) [inline eng/pipelines/corefx-base.yml]: Why is this declaration needed? If we're going to declare a variable to store a parameter value, then we should use the variable elsewhere within the file.

Comment (mikem8361) [reply eng/pipelines/corefx-base.yml]: "parameter.targetOS" didn't work  when I used it directly in the "publish test assets..." step. The only way I got it to work is to create this variable.

**Comment by mikem8361** [comment]
@safern Is this PR ready to merge?

File: eng/pipelines/corefx-base.yml (hunk 6 above)
Comment (safern) [reply eng/pipelines/corefx-base.yml]: Maybe syntax was incorrect? We do it like that elsewhere -- i.e: https://github.com/dotnet/corefx/blob/master/eng/pipelines/helix.yml#L22

**Comment by safern** [comment]
I'd like this feedback to be resolved:

https://github.com/dotnet/corefx/pull/35473#discussion_r258752668

I don't want to introduce extra variables that aren't needed, just for readability.

**Comment by mikem8361** [comment]
I'll remove the _targetOS and try to use `parameters.targetOS` directory in the step. I'll have to do another official build to test it.

**Comment by safern** [comment]
> I'll remove the _targetOS and try to use `parameters.targetOS` directly in the step. I'll have to do another official build to test it.

Ok, sorry for the pickiness, yaml file is already hard to read, don't want to over complicate it with this kind of small changes.

**Comment by mikem8361** [comment]
No problem and I think my problem before was that I need to use `${{ parameters.targetOS }}` instead of just `$(parameters.targetOS)`. It is a good yaml learning experience.

**Comment by safern** [comment]
Yep, for parameters you need to use ${{ ... }} syntax, 😄 -- thanks for fixing.

File: eng/pipelines/corefx-base.yml [hunk 7]
```diff
@@ -55,6 +59,9 @@ jobs:
     - template: ../common/templates/job/job.yml
       parameters:
         variables:
+        - name: _dotnetFeedUrl
+          value: https://dotnetfeed.blob.core.windows.net/dotnet-core/index.json
```
Comment (safern) [inline eng/pipelines/corefx-base.yml]: Sorry for going back and forth on this, for variable names, we use the following syntax:

- variableName: value -- `- _dotnetFeedUrl: https://dotnetfeed.blob.core.windows.net/dotnet-core/index.json`

Comment (safern) [reply eng/pipelines/corefx-base.yml]: No need, to do an official build for this, if the PR builds compile and run correctly, then the syntax is correct.

Comment (safern) [reply eng/pipelines/corefx-base.yml]: Also, since this is just used in official builds you could move it down to the if statement where you declare 
`-group: DotNet-Blob-Feed` 

https://github.com/dotnet/corefx/pull/35473/files#diff-4aae51df678fdaa7a5e0769b0593310aR90

File: eng/publish.proj [hunk 8]
```diff
@@ -104,4 +104,56 @@
...
+    </PropertyGroup>
+
+    <ItemGroup>
+      <_ItemsToPush Remove="@(_ItemsToPush)" />
+      <_ItemsToPush Include="$(FilesToPublishPattern)" />
+      <_ItemsToPush>
+        <RelativeBlobPath>$(_FileRelativePathBase)$([System.String]::Copy('%(RecursiveDir)%(Filename)%(Extension)').Replace('\' ,'/'))</RelativeBlobPath>
```
Comment (safern) [inline eng/publish.proj]: Is this trying to do what MSBuild Normalize path do?

https://github.com/dotnet/arcade/blob/beadc52f70da2965fbab941e9b4daa5ddf3e3291/src/Microsoft.DotNet.CoreFxTesting/build/core/Core.targets#L54

```suggestion
        <RelativeBlobPath>$([MSBuild]::NormalizePath('$(_FileRelativePathBase)', '%(RecursiveDir)', '%(Filename)%(Extension)'))</RelativeBlobPath>
```

Comment (mikem8361) [reply eng/publish.proj]: I stole that line from src/uploadtests.proj. I'll replace it with your suggestion.

**Comment by mikem8361** [comment]
Yes, it may be useful to be published as part of the build artifacts. I'll add it.

File: eng/pipelines/corefx-base.yml [hunk 9]
```diff
@@ -79,6 +84,10 @@ jobs:
         - ${{ if eq(job.submitToHelix, 'true') }}:
           - group: DotNet-HelixApi-Access
 
+        - ${{ if eq(parameters.isOfficialBuild, 'true') }}:
+          - group: DotNet-Blob-Feed
+          - _dotnetFeedUrl: https://dotnetfeed.blob.core.windows.net/dotnet-core/index.json
```
Comment (safern) [inline eng/pipelines/corefx-base.yml]: Just as an FYI that this syntax that I suggested is not officially supported by azure devops for variables, but since we use arcade's template job.yml -- it parses the variables that come as objects and transforms them to the name-variable syntax that you had before:
```
- name: <name>
   value: <variable value>
```

https://github.com/dotnet/arcade/blob/master/eng/common/templates/job/job.yml#L103-L109

File: eng/pipelines/corefx-base.yml [hunk 10]
```diff
@@ -173,6 +182,34 @@ jobs:
...
+                      /p:ManifestCommit=$(Build.SourceVersion)
+                      /p:ManifestRepoUri=$(Build.Repository.Uri)
+                displayName: Publish test assets to dotnet-core feed
+                condition: and(succeeded(), eq(variables['_publishTests'], 'true'))
+
+              - task: PublishBuildArtifacts@1
+                displayName: Publish test asset manifest to artifacts container
+                inputs:
+                  pathToPublish: $(Build.SourcesDirectory)/artifacts/AssetManifests
+                  artifactName: manifests
```
Comment (safern) [inline eng/pipelines/corefx-base.yml]: Could you named this: `BuildAssetsManifest` please? We're publishing another assets after we publish packages, so that all the .xml manifests are in the same container:

https://github.com/dotnet/corefx/blob/master/eng/pipelines/publish.yml#L73

Also, from you official build I noticed the .xml file name is not corefx-test-assets.xml and it will clash with the asset manifest that we produce later when publishing packages. Could you please make it be called corefx-test-assets.xml ?

You just need to pass AssetManifestName down to publish.proj:

https://github.com/dotnet/corefx/blob/master/eng/publish.proj#L11

File: eng/publish.proj [hunk 11]
```diff
@@ -104,4 +104,56 @@
...
+    <PushToBlobFeed ExpectedFeedUrl="$(ExpectedFeedUrl)"
+                    AccountKey="$(AccountKey)"
+                    ItemsToPush="@(_ManifestToPush)"
+                    PublishFlatContainer="true"
+                    ManifestBuildId="$(ManifestBuildId)"
+                    ManifestBranch="$(ManifestBranch)"
+                    ManifestCommit="$(ManifestCommit)"
+                    ManifestBuildData="$(ManifestBuildData)"
+                    ManifestRepoUri="$(ManifestRepoUri)"
+                    AssetManifestPath="$(ArtifactsDir)AssetManifests\ManifestUpload.xml" />
```
Comment (safern) [inline eng/publish.proj]: Nit: add an extra msbuild variable at the top to define `$(ArtifactsDir)\AssetManifests` since now it is used in multiple places. So that way here you will just do:

`$(AssetManifestDir)\ManifestUpload.xml`

File: eng/pipelines/corefx-base.yml (hunk 7 above)
Comment (mikem8361) [reply eng/pipelines/corefx-base.yml]: FYI, publish.yml uses the value/name syntax for all of it's variables.  It is where I copied _dotnetFeedUrl in the first place.

File: eng/pipelines/corefx-base.yml (hunk 10 above)
Comment (mikem8361) [reply eng/pipelines/corefx-base.yml]: I don't think we need the publish artifacts step in corefx-base.yml because the one in publish.yml will publish them. 

I'm passing the corefx-test-assets.xml asset manifest name to publish.proj.

Comment (mikem8361) [reply eng/pipelines/corefx-base.yml]: Never mind about the removing the corefx-test-assets.xml publish step to the build artifacts.  It needs to be there and I need to include the build leg name in the artifactName. Like "Windows_NT_x64_Release" or "Linux_x64_Debug" to put it along side the build logs.

File: eng/pipelines/corefx-base.yml (hunk 7 above)
Comment (safern) [reply eng/pipelines/corefx-base.yml]: publish.yml uses that syntax because it doesn't use the Job.yml template from arcade, so the cool `-key: value` syntax is not available for that yml file 😄

File: eng/pipelines/corefx-base.yml (hunk 10 above)
Comment (safern) [reply eng/pipelines/corefx-base.yml]: > I don't think we need the publish artifacts step in corefx-base.yml because the one in publish.yml will publish them.

It will not publish them, because that step runs in a different machine, different environment, so the files will not be available in the publish step. We can talk about this offline if it helps.
